### PR TITLE
Fast and compile friendly text generation

### PR DESCRIPTION
**Pull request by carmocca**
# generate/base.py

Eager (before this PR): 39.97 tok/sec, 13.77 GB
Eager: 41.32 tok/sec, 13.76 GB
Compile: 91.16 tok/sec, 14.01 GB

`python generate/base.py --checkpoint_dir checkpoints/openlm-research/open_llama_7b --precision bf16-true --max_new_tokens 512 --num_samples 10`
Using the last sample generated

# chat/base.py

Eager (before this PR): ~40 tok/sec
Eager: ~41 tok/sec
Compile: ~65 tok/sec

`python chat/base.py --checkpoint_dir checkpoints/openlm-research/open_llama_7b --precision bf16-true --compile true`

# Environment

```python
pytorch-triton           2.1.0+6e4932cda8
torch                    2.2.0.dev20231120+cu121
GPU 0: NVIDIA A100-SXM4-40GB
CUDA runtime version: 12.1.105
```

### CODE REVIEW

File: generate/adapter.py [hunk 1]
```diff
@@ -104,23 +117,9 @@ def main(
     model.load_state_dict(checkpoint)
     fabric.print(f"Time to load the model weights: {time.perf_counter() - t0:.02f} seconds.", file=sys.stderr)
 
-    model.eval()
     model = fabric.setup(model)
 
-    tokenizer = Tokenizer(checkpoint_dir)
```
Comment (awaelchli) [inline generate/adapter.py:110]: Curious what's the impact of moving these blocks of code up?

File: generate/base.py [hunk 2]
```diff
@@ -106,6 +110,7 @@ def main(
     strategy: str = "auto",
     devices: int = 1,
     precision: Optional[str] = None,
+    compile: bool = False,
```
Comment (awaelchli) [inline generate/base.py:113]: We expose this as optional because of the impact during warm up, right?

File: generate/base.py [hunk 3]
```diff
@@ -52,50 +68,38 @@ def generate(
...
-        if idx_next == eos_id:
-            return idx[:input_pos]  # include the EOS token
-
-    return idx
+    device = prompt.device
+    tokens = [prompt]
+    input_pos = torch.tensor([T], device=device)
+    token = next_token(
+        model, torch.arange(0, T, device=device), prompt.view(1, -1), temperature=temperature, top_k=top_k
+    ).clone()
```
Comment (awaelchli) [inline generate/base.py:90]: I'm curious why is iteration 1 done separate from the loop?

**Review by awaelchli — APPROVED** [review]

File: generate/adapter.py (hunk 1 above)
Comment (carmocca) [reply generate/adapter.py:110]: No impact speed-wise. I originally was creating the kv cache inside `init_module` since I don't need to do it in the loop anymore but `empty_init` messed it up with fsdp so I went back to using `init_tensor`.

TLDR: this change is not necessary anymore, I could revert it if you prefer the previous order

File: generate/base.py (hunk 3 above)
Comment (carmocca) [reply generate/base.py:90]: This is in the prefill stage of the kv cache so the input shapes are different. Originally I was compiling the prefill and next_token calls separately but later experiments showed that I didn't have much of an impact. I kept it this way because I want to allocate the `input_pos` tensors asap instead of having only one that is modified in the loop.

File: generate/base.py (hunk 2 above)
Comment (carmocca) [reply generate/base.py:113]: Yes, warmup can take a long time (1-2 min).
